### PR TITLE
Refactor KeyCondition and key analysis

### PR DESCRIPTION
**Pull request by amosbird**
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Refactor KeyCondition and key analysis to improve PartitionPruner and trivial count optimization. This is separated from #60463 .

### CODE REVIEW

**Comment by robot-clickhouse** [comment]
*This is an automated comment for commit 79d507697c06009f3f0c62d2030563ae558e18f2 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/61459/79d507697c06009f3f0c62d2030563ae558e18f2/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

File: src/Storages/MergeTree/BoolMask.h [hunk 1]
```diff
@@ -1,37 +1,59 @@
...
-    BoolMask operator !() const
+    BoolMask operator&(const BoolMask & m) const { return {can_be_true && m.can_be_true, can_be_false || m.can_be_false}; }
+    BoolMask operator|(const BoolMask & m) const { return {can_be_true || m.can_be_true, can_be_false && m.can_be_false}; }
+    BoolMask operator!() const { return {can_be_false, can_be_true}; }
+
+    bool operator==(const BoolMask & other) const { return can_be_true == other.can_be_true && can_be_false == other.can_be_false; }
+
+    /// Check if mask is no longer changeable under BoolMask::combine.
+    /// We use this condition to early-exit KeyConditions::checkInRange methods.
+    bool isComplete(const BoolMask & initial_mask) const
```
Comment (KochetovNicolai) [inline src/Storages/MergeTree/BoolMask.h:22]: initial_mask is a bit confusing here. What values do we expect from it?

Maybe we can do `this->combine(initial_mask).isComplete()` check on the call side?

File: src/Storages/MergeTree/KeyCondition.cpp [hunk 2]
```diff
@@ -2274,25 +2295,24 @@ static BoolMask forAnyHyperrectangle(
...
-    result = result | callback(hyperrectangle);
+    auto result = BoolMask::combine(initial_mask, callback(hyperrectangle));
 
     /// There are several early-exit conditions (like the one below) hereinafter.
-    /// They are important; in particular, if initial_mask == BoolMask::consider_only_can_be_true
-    /// (which happens when this routine is called from KeyCondition::mayBeTrueXXX),
-    /// they provide significant speedup, which may be observed on merge_tree_huge_pk performance test.
-    if (result.isComplete())
+    /// They provide significant speedup, which may be observed on merge_tree_huge_pk performance test.
+    if (result.isComplete(initial_mask))
```
Comment (KochetovNicolai) [inline src/Storages/MergeTree/KeyCondition.cpp:2302]: This change I don't fully understand.
First of all, the logic has changed. Previously we used `result | callback(hyperrectangle)`, and now we use `BoolMask::combine(initial_mask, callback(hyperrectangle))`, which gives a different result.

Alos now `result.isComplete` takes `initial_mask` again as an argument. Is it needed here?

Comment (KochetovNicolai) [reply src/Storages/MergeTree/KeyCondition.cpp:2302]: Or, probably, there was a bug previously?
Than, let's make a separate fix and backport it.

**Review by KochetovNicolai — COMMENTED** [review]
Overall looks good.

File: src/Storages/MergeTree/BoolMask.h (hunk 1 above)
Comment (amosbird) [reply src/Storages/MergeTree/BoolMask.h]: initial_mask argument is removed

**Comment by amosbird** [comment]
> prewhere_profile_events

Also observed as flaky in #64255, #61112

> test_checking_s3_blobs_paranoid/test.py::test_query_is_canceled_with_inf_retries

S3 write failure.
```
E           helpers.client.QueryRuntimeException: Client failed! Return code: 243, stderr: Received exception from server (version 24.6.1):
E           Code: 499. DB::Exception: Received from 172.16.3.6:9000. DB::Exception: Poco::Exception. Code: 1000, e.code() = 111, Connection refused (version 24.6.1.2668). Stack trace:
E           
E           0. DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool) @ 0x000000000bc2f768
E           1. DB::S3Exception::S3Exception(String const&, Aws::S3::S3Errors) @ 0x000000000e757374
E           2. void std::__function::__policy_invoker<void ()>::__call_impl<std::__function::__default_alloc_func<DB::WriteBufferFromS3::writePart(DB::WriteBufferFromS3::PartData&&)::$_0, void ()>>(std::__function::__policy_storage const*) @ 0x000000000e7f0424
E           3. void std::__function::__policy_invoker<void ()>::__call_impl<std::__function::__default_alloc_func<DB::TaskTracker::add(std::function<void ()>&&)::$_0, void ()>>(std::__function::__policy_storage const*) @ 0x000000000e7f47cc
E           4. std::__packaged_task_func<std::function<std::future<void> (std::function<void ()>&&, Priority)> DB::threadPoolCallbackRunnerUnsafe<void, std::function<void ()>>(ThreadPoolImpl<ThreadFromGlobalPoolImpl<false, true>>&, String const&)::'lambda'(std::function<void ()>&&, Priority)::operator()(std::function<void ()>&&, Priority)::'lambda'(), std::allocator<std::function<std::future<void> (std::function<void ()>&&, Priority)> DB::threadPoolCallbackRunnerUnsafe<void, std::function<void ()>>(ThreadPoolImpl<ThreadFromGlobalPoolImpl<false, true>>&, String const&)::'lambda'(std::function<void ()>&&, Priority)::operator()(std::function<void ()>&&, Priority)::'lambda'()>, void ()>::operator()() @ 0x000000000e756254
E           5. std::packaged_task<void ()>::operator()() @ 0x000000000d5460b8
E           6. ThreadPoolImpl<ThreadFromGlobalPoolImpl<false, true>>::worker(std::__list_iterator<ThreadFromGlobalPoolImpl<false, true>, void*>) @ 0x000000000bccb624
E           7. void std::__function::__policy_invoker<void ()>::__call_impl<std::__function::__default_alloc_func<ThreadFromGlobalPoolImpl<false, true>::ThreadFromGlobalPoolImpl<void ThreadPoolImpl<ThreadFromGlobalPoolImpl<false, true>>::scheduleImpl<void>(std::function<void ()>, Priority, std::optional<unsigned long>, bool)::'lambda0'()>(void&&)::'lambda'(), void ()>>(std::__function::__policy_storage const*) @ 0x000000000bccf190
E           8. void* std::__thread_proxy[abi:v15000]<std::tuple<std::unique_ptr<std::__thread_struct, std::default_delete<std::__thread_struct>>, void ThreadPoolImpl<std::thread>::scheduleImpl<void>(std::function<void ()>, Priority, std::optional<unsigned long>, bool)::'lambda0'()>>(void*) @ 0x000000000bccde40
E           9. ? @ 0x000000000007d5c8
E           10. ? @ 0x00000000000e5edc
E           . (S3_ERROR)
E           (query: KILL QUERY WHERE query_id = 'TEST_QUERY_IS_CANCELED_WITH_INF_RETRIES' ASYNC)
```

> test_replicated_database/test.py::test_alter_attach[ReplicatedMergeTree]	

Also observed flaky in #64694

**Comment by amosbird** [comment]
Stateless tests (coverage) somehow fails to run.

**Review by KochetovNicolai — APPROVED** [review]

**Pull request merged**